### PR TITLE
🐳Add a Dockerfile and Webserver listening Command Line Parameter

### PR DESCRIPTION
**Pull request by dominikkaminski**
**Changes:**

1. Add Multistage Dockerfile
2. Add .dockerignore in order to reduce build context and image size
3. Add How-to-use Docker image in README.md
4. Add webserver host listening parameter. Now it is possible to change listener via command line parameter
5. Add host parameter to README.md

## How can others test the changes?

**Docker**

```shell
docker build --build-arg node_version=10-alpine \
             -t bpmn:v0.1 .
```

```shell
docker build -t bpmn:v0.1 .
```

**Webserver listening parameter**

```shell
npm start -- --host=0.0.0.0
```
## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -141,6 +151,35 @@ Die Releases des BPMN-Studios lassen sich alternativ auch
...
+
+Der Container lässt sich mit folgendem Befehl starten:
+
+```shell
+docker run -p 17290:17290 bpmn:v0.1
+```
+
+Anschließend lässt sich das BPMN-Studio unter URL `http://localhost:17290` aufrufen.
+
+#### Container bauen
```
Comment (Vyperus) [inline README.md]: Sollte ich den Container nicht erst bauen und dann starten? Pass bitte die Reihenfolge an

File: bin/bpmn-studio.js [hunk 2]
```diff
@@ -49,3 +51,15 @@ function _applicationPortIsValid(port) {
 
   return true;
 }
+
+function _applicationHostIstValid(host) {
```
Comment (Vyperus) [inline bin/bpmn-studio.js]: `_applicationHostIsValid`

File: README.md [hunk 3]
```diff
@@ -141,6 +151,35 @@ Die Releases des BPMN-Studios lassen sich alternativ auch
...
+
+Das Image lässt sich wie folgt bauen:
+
+```shell
+docker build -t bpmn:v0.1 .
+```
+
+Es ist möglich, das base image, sowie die Paketversionen anzupassen:
```
Comment (Vyperus) [inline README.md]: Diesen Abschnitt bitte auch als solchen Auszeichen (mit Überschrift). 
Und dann die möglichen Optionen/Parameter genauer beschreiben.

File: README.md [hunk 4]
```diff
@@ -141,6 +151,35 @@ Die Releases des BPMN-Studios lassen sich alternativ auch
...
+
+* `node_version`: Base image version mit NodeJS und Alpine Linux
+
+```shell
+docker build --build-arg node_version=10-alpine \
+             -t bpmn:v0.1 .
```
Comment (cmg-dev) [inline README.md]: Ich bitte immer darum in Skripten und in Doku, die cli-parameter aus zu schreiben:

```
docker build \
      --build-arg node_version=10-alpine \
      --tag bpmn:v0.1 .
```

File: bin/bpmn-studio.js [hunk 5]
```diff
@@ -5,19 +5,21 @@ const argv = require('minimist')(process.argv.slice(2));
 const pushserve = require('pushserve');
 
 const defaultPort = 17290;
+const defaultHost = '127.0.0.1';
 const portUsed = _applicationPortIsValid(argv.port) ? argv.port : defaultPort;
+const hostUsed = _applicationHostIstValid(argv.host) ? argv.host : defaultHost;
```
Comment (cmg-dev) [inline bin/bpmn-studio.js]: 👍

File: bin/bpmn-studio.js [hunk 6]
```diff
@@ -49,3 +51,15 @@ function _applicationPortIsValid(port) {
 
   return true;
 }
+
+function _applicationHostIstValid(host) {
+  if (host === null || host === undefined) {
+    return false;
+  }
+  host_array = host.split('.');
+  if (host_array.length !== 4) {
```
Comment (cmg-dev) [inline bin/bpmn-studio.js]: Nur für die Zukunft:

Ich würde IMMER die business-Regel extrahieren, die hier implizit steht:

```js
(...)
host_array = host.split('.');

const addressHasNotFourOctetts = host_array.length !== 4;
if (addressHasNotFourOctetts) {
(...)
```

Das ist nämlich, was du hier testest, was aber nicht jedem klar ist. So hilft das aber jedem der drauf guckt.

**Pull request merged**